src/lib/scheduler/operator_task.cpp

```diff
@@ -23,33 +23,32 @@ std::string OperatorTask::description() const {
 const std::vector<std::shared_ptr<OperatorTask>> OperatorTask::make_tasks_from_operator(
     std::shared_ptr<AbstractOperator> op) {
   std::vector<std::shared_ptr<OperatorTask>> tasks;
-  OperatorTask::_add_tasks_from_operator(op, tasks);
+  std::unordered_map<std::shared_ptr<AbstractOperator>, std::shared_ptr<OperatorTask>> task_by_op;
+  OperatorTask::_add_tasks_from_operator(op, tasks, task_by_op);
   return tasks;
 }
 
 std::shared_ptr<OperatorTask> OperatorTask::_add_tasks_from_operator(
-    std::shared_ptr<AbstractOperator> op, std::vector<std::shared_ptr<OperatorTask>>& tasks) {
-  std::shared_ptr<OperatorTask> task = op->operator_task();
-  bool add_task = false;
-  if (!task) {
-    // We need to make sure that we don't create two tasks for the same operator. This could happen if the same
-    // operator is used as input for two other operators.
-    task = std::make_shared<OperatorTask>(op);
-    op->set_operator_task(task);
-    add_task = true;
-  }
+    std::shared_ptr<AbstractOperator> op, std::vector<std::shared_ptr<OperatorTask>>& tasks,
+    std::unordered_map<std::shared_ptr<AbstractOperator>, std::shared_ptr<OperatorTask>>& task_by_op) {
+  const auto task_by_op_it = task_by_op.find(op);
+  if (task_by_op_it != task_by_op.end()) return task_by_op_it->second;
+
+  const auto task = std::make_shared<OperatorTask>(op);
+  task_by_op.emplace(op, task);
```

Comment:
Why does this become necessary? Adding this map adds some lookup costs - couldn't we go by the `operator_task()` that we had before and simply clear it when we recreate an operator?

Reply:
Having the Operator know that there is such a concept as the OperatorTask always seemed dirty to me. Generally, objects with knowledge about their enclosing object are anti patterns. This even looks like a circular reference, now that I look at it. (which we could of course work around) (Humongous amounts of memory, anyone? https://github.com/hyrise/hyrise/issues/656)

Your call, but I'd like to keep it this way.

Reply:
Yes, this is circular, but could easily be broken with a `weak_ptr`.

I believe that there was another reason why we introduced it, which was to make sure that no operator was scheduled twice: 92058cf02e67ef42fc053342b9637097cf7ac67e